codec-http2/src/main/java/io/netty/handler/codec/http2/StreamBufferingEncoder.java

```diff
@@ -0,0 +1,347 @@
+                padding, endOfStream, promise));
+        return promise;
+    }
+
+    @Override
+    public ChannelFuture writeRstStream(ChannelHandlerContext ctx, int streamId, long errorCode,
+                                        ChannelPromise promise) {
+        if (existingStream(streamId)) {
+            return super.writeRstStream(ctx, streamId, errorCode, promise);
+        }
+        // Since the delegate doesn't know about any buffered streams we have to handle cancellation
+        // of the promises and releasing of the ByteBufs here.
+        PendingStream stream = pendingStreams.remove(streamId);
+        if (stream != null) {
+            // Sending a RST_STREAM to a buffered stream will succeed the promise of all frames
+            // associated with the stream, as sending a RST_STREAM means that someone "doesn't care"
+            // about the stream anymore and thus there is not point in failing the promises and invoking
+            // error handling routines.
+            stream.close(null);
+            promise.setSuccess();
+        } else {
+            promise.setFailure(
+                    connectionError(PROTOCOL_ERROR, "Stream does not exist %d", streamId));
+        }
+        return promise;
+    }
+
+    @Override
+    public ChannelFuture writeData(ChannelHandlerContext ctx, int streamId, ByteBuf data,
+                                   int padding,
+                                   boolean endOfStream, ChannelPromise promise) {
+        if (existingStream(streamId)) {
+            return super.writeData(ctx, streamId, data, padding, endOfStream, promise);
+        }
+        PendingStream pendingStream = pendingStreams.get(streamId);
+        if (pendingStream != null) {
+            pendingStream.frames.add(new DataFrame(data, padding, endOfStream, promise));
+        } else {
+            ReferenceCountUtil.safeRelease(data);
+            promise.setFailure(
+                    connectionError(PROTOCOL_ERROR, "Stream does not exist %d", streamId));
+        }
+        return promise;
+    }
+
+    @Override
+    public ChannelFuture writeSettingsAck(ChannelHandlerContext ctx, ChannelPromise promise) {
+        receivedSettings = true;
+        ChannelFuture future = super.writeSettingsAck(ctx, promise);
+        // After having received a SETTINGS frame, the maximum number of concurrent streams
+        // might have changed. So try to create some buffered streams.
+        tryCreatePendingStreams();
+        return future;
+    }
+
+    @Override
+    public void close() {
+        super.close();
+        cancelPendingStreams();
+    }
+
+    private void tryCreatePendingStreams() {
+        while (!pendingStreams.isEmpty() && canCreateStream()) {
+            Map.Entry<Integer, PendingStream> entry = pendingStreams.pollFirstEntry();
+            PendingStream pendingStream = entry.getValue();
+            pendingStream.sendFrames();
+            largestCreatedStreamId = pendingStream.streamId;
```

Comment:
@Scottmitch I made this so that we can decide with one comparison whether to delegate to the actual encoder or whether to buffer. One is not supposed to be able to enqueue a data frame before a headers frame? The call to `writeData` should fail if no headers frame has been written before. Otherwise that's a bug.